src/Compilers/Core/Portable/Operations/CommonConversion.cs

```diff
@@ -19,7 +19,14 @@ private enum ConversionKind
             Exists = 0b0001,
             IsIdentity = 0b0010,
             IsNumeric = 0b0100,
-            IsReference = 0b1000
+            IsReference = 0b1000,
+
+            /// <summary>
+            /// Conversion is not created by language specific analysis. 
+            /// No language specific information can be retrieved for this conversion
+            /// from language specific APIs.
+            /// </summary>
+            IsLanguageAgnostic = 0b10000,
```

Comment:
Can you please update the extension methods to detect these conversions and have them throw an appropriate exception? Right now, I believe they'll just throw a very unhelpful ClassCastException.

Reply:
> Can you please update the extension methods to detect these conversions and have them throw an appropriate exception? Right now, I believe they'll just throw a very unhelpful ClassCastException.

This is how the behavior is documented. 

---
In reply to: [178665399](https://github.com/dotnet/roslyn/pull/25828#discussion_r178665399) [](ancestors = 178665399)

Reply:
Currently, we throw an exception with a helpful error message for the wrong language. We should do the same for a language agnostic conversion as well.

---
In reply to: [178665972](https://github.com/dotnet/roslyn/pull/25828#discussion_r178665972) [](ancestors = 178665972,178665399)

Reply:
At the very least, we need to add some tests for the scenario.

---
In reply to: [178666871](https://github.com/dotnet/roslyn/pull/25828#discussion_r178666871) [](ancestors = 178666871,178665972,178665399)

Reply:
Currently, we throw an exception with a helpful error message for the wrong language.

I don't see this, this is what is documented and we do not check that that it is safe to cast anything we just cast and the behavior is inline with the documentation on the methods:
```
<exception cref="InvalidCastException">If the <see cref="IConversionOperation"/> was not created from Visual Basic code.</exception>
/// <exception cref="InvalidCastException">If the <see cref="IConversionOperation"/> was not created from CSharp code.</exception>
```
 



---
In reply to: [178666871](https://github.com/dotnet/roslyn/pull/25828#discussion_r178666871) [](ancestors = 178666871,178665972,178665399)

Reply:
> At the very least, we need to add some tests for the scenario.

There are tests for this scenario. They all go through VisitConversion.VisitConversion

---
In reply to: [178668850](https://github.com/dotnet/roslyn/pull/25828#discussion_r178668850) [](ancestors = 178668850,178666871,178665972,178665399)

Reply:
[https://github.com/dotnet/roslyn/blob/features/dataflow/src/Compilers/CSharp/Portable/CSharpExtensions.cs#L703](https://github.com/dotnet/roslyn/blob/features/dataflow/src/Compilers/CSharp/Portable/CSharpExtensions.cs#L703)

We check to see if the conversion is from C# first. I think we should do the same with a language agnostic conversion.

---
In reply to: [178668850](https://github.com/dotnet/roslyn/pull/25828#discussion_r178668850) [](ancestors = 178668850,178666871,178665972,178665399)

Reply:
This is checking the language for the syntax and has nothing to do with determining where the conversion came from.

---
In reply to: [178669374](https://github.com/dotnet/roslyn/pull/25828#discussion_r178669374) [](ancestors = 178669374,178668850,178666871,178665972,178665399)

Reply:
If you think that method should not throw InvalidCastException, feel free to open an issue. The way I see it this is the current design which is in agreement with the current implementation

---
In reply to: [178669740](https://github.com/dotnet/roslyn/pull/25828#discussion_r178669740) [](ancestors = 178669740,178669374,178668850,178666871,178665972,178665399)